### ISSUE TITLE
Cannot use "import" in `meteor shell` via STDIN

### ISSUE THREAD

**Issue by atd** [meteor/meteor#8823]
While using `import` in a normal `meteor shell` works

```
$ meteor shell

Welcome to the server-side interactive shell!

Tab completion is enabled for global variables.

Type .reload to restart the server and the shell.
Type .exit to disconnect from the server and leave the shell.
Type .help for additional help.

> import { Success } from "./imports/collections";
> Success
'success!'
> Shell exiting...
```

The same command using STDIN throws a `SyntaxError`

```
$ echo 'import { Success } from "./imports/collections"' | meteor shell
SyntaxError: Unexpected token, expected ( (1:8)
```

Seen in both meteor `1.4.4.3` and `1.5`

**Comment by atd** [comment]
The issue can be reproduced using https://github.com/atd/meteor-issue-8823

**Comment by atd** [comment]
So I guess the cause of the issue is this line

https://github.com/meteor/meteor/blob/7bd0382ad96c60a24f498e3740e2ad8d20da8709/tools/static-assets/server/shell-server.js#L175

I think I found the solution to the [comma syntax mistery](https://forums.meteor.com/t/how-to-remote-control-the-meteor-shell-from-bash/32704)

The funny thing is that you can build a "meteor shell command injection" and bypass the parenthesis restriction:

```
$ echo '""); import { Success } from "./imports/collections"; (Success' | meteor shell
"success!"
```

So why the parenthesis wrapping when just sending multiple commands seems to work?

**Comment by benjamn** [comment]
If I recall correctly, the reason is so that the code is treated as an expression, so that it has a value that can be echoed back. Would you mind playing with the `shell-server.js` implementation to see what happens if you don't force the expression with parentheses? We'd be happy to consider a pull request!

**Comment by atd** [comment]
It seems things work like before: the output shows the value of the last expression

```
$ echo 'var a = 1' | meteor shell                                                    
undefined
$ echo 'var a = 1; a + 1' | meteor shell                                             
2
$ echo 'console.log("hi"); var a = 1; a + 1' | meteor shell                        
2
```
(the server outputs `I20170621-12:19:27.953(2)? hi`)

The variables are saved across executions:

```
$ echo 'a + 1' | meteor shell                                             
2
```

There is an additional problem through: `require` and `module` are missing. They are added to the `repl` context [here](https://github.com/meteor/meteor/blob/7bd0382ad96c60a24f498e3740e2ad8d20da8709/tools/static-assets/server/shell-server.js#L261), but they seem not to be present when the [vm script is run](https://github.com/meteor/meteor/blob/7bd0382ad96c60a24f498e3740e2ad8d20da8709/tools/static-assets/server/shell-server.js#L363), (through they should be present in the context?)

```
$ echo 'import { Success } from "./imports/collections"' | meteor shell              
ReferenceError: module is not defined
```

However, they present once the import has been run in a repl

```
$ meteor shell 

Welcome to the server-side interactive shell!

Tab completion is enabled for global variables.

Type .reload to restart the server and the shell.
Type .exit to disconnect from the server and leave the shell.
Type .help for additional help.

> import { Success } from './imports/collections';
> Success
'success!'
> Shell exiting...

$ echo 'import { Ok } from "./imports/collections"; Ok' | meteor shell
"ok"
```


I'd be happy to prepare a pull request.

**Comment by atd** [comment]
An explanation of the `module` missing issue

https://github.com/nodejs/node-v0.x-archive/issues/9211#issuecomment-134412415

**Comment by benjamn** [comment]
@atd Thanks for all the digging! I would just point out that we're already implementing our own REPL server, so we have a lot of control over what variables are exposed to it, whereas I think that issue is concerning the default Node REPL. In other words, we should be able to make this work. Please do submit a PR for any changes you'd like to see, even if it's still a work in progress, whenever you get a chance.

**Comment by atd** [comment]
The issue is that `repl` is not started when running a script, [this line is not reached](https://github.com/meteor/meteor/blob/7bd0382ad96c60a24f498e3740e2ad8d20da8709/tools/static-assets/server/shell-server.js#L213) and therefore [`require` and `module` aren't added to the context](https://github.com/meteor/meteor/blob/7bd0382ad96c60a24f498e3740e2ad8d20da8709/tools/static-assets/server/shell-server.js#L261)

I have prepare a quick hack to make them available: https://github.com/meteor/meteor/pull/8833
Note that `require` and `module` end up being added to `this` (or `repl`) because it is the same object used in `repl.context`. In any case, I am open to comments and improvements on the approach.

Finally, I am wondering why you don't just pipe the streams in https://github.com/meteor/meteor/blob/devel/tools/shell-client.js I think it would make `meteor shell` more powerful

**Comment by abernix** [comment]
@atd Thanks for looking into this! I think your links are missing the line number anchors (e.g. `#L15`) so I didn't immediately see what you were referring to.

It's worth pointing out that a lot of the REPL implementation was done before much of the [`repl`](https://nodejs.org/api/repl.html#repl_repl) functionality in Node.js was fully exposed.  Some parts (for example, [this code](https://github.com/meteor/meteor/blob/2a3a01acb36620036e6dd33ca05bcb6313eef8a4/packages/shell-server/shell-server.js#L270-L294) catching to get the previously not-`export`ed `RecoverableError` error class) are simply not necessary anymore.

There are new examples in the docs (linked above) so maybe worth re-evaluating those.

**Comment by atd** [comment]
Ok, so tests exposed the reason of the parentheses:

```
$ echo "{server:Meteor.isServer}" | /home/atd/dev/meteor/meteor shell
true
```

Sorry @abernix which part needs clarification? The reason of why [`require` and `module` need to be added to context](https://github.com/atd/meteor/blob/c1e3aea26f4aafc3b91f5035382705f78be3d10f/packages/shell-server/shell-server.js#L448-L451), the streams pipe question, or both?

**Comment by abernix** [comment]
@atd I was just pointing out that some of the context you were referring to via`she'll-client.js` links were pointing to the whole file, not specific lines (you referred to "here" as if you were pointing to a specific line).   So yes, all of the above, but specificaly I think the stream question. ;)

**Comment by atd** [comment]
Ok, so I was wondering why `shell-client.js` [handles the script case in a different way](https://github.com/meteor/meteor/blob/devel/tools/shell-client.js#L124-L126) by just [sending a `evaluateAndExit` command](https://github.com/meteor/meteor/blob/devel/tools/shell-client.js#L92-L103) and it just doesn't [pipe the `process.stdin` like in the tty case](https://github.com/meteor/meteor/blob/devel/tools/shell-client.js#L146).

I think that it would allow process piping at the OS shell level and do more powerful things. Besides using the same code in `shell-server.js`

**Comment by benjamn** [comment]
Should be fixed in `shell-server@0.2.4`, officially released with Meteor 1.5.1.
